src/controller/CHIPDeviceController.cpp

```diff
@@ -1494,12 +1494,23 @@ void DeviceCommissioner::OnDeviceConnectionFailureFn(void * context, PeerId peer
 {
     // CASE session establishment failed.
     DeviceCommissioner * commissioner = static_cast<DeviceCommissioner *>(context);
+
     ChipLogProgress(Controller, "Device connection failed. Error %s", ErrorStr(error));
     VerifyOrReturn(commissioner != nullptr,
                    ChipLogProgress(Controller, "Device connection failure callback with null context. Ignoring"));
     VerifyOrReturn(commissioner->mPairingDelegate != nullptr,
                    ChipLogProgress(Controller, "Device connection failure callback with null pairing delegate. Ignoring"));
 
+    //
+    // If a device is being commissioned currently and it is the very same device that we just failed to establish CASE with,
+    // we need to clean it up to prevent a dangling CommissioneeDeviceProxy object.
+    //
+    if (commissioner->mDeviceBeingCommissioned != nullptr && commissioner->mDeviceBeingCommissioned->GetPeerId() == peerId)
+    {
+        commissioner->ReleaseCommissioneeDevice(commissioner->mDeviceBeingCommissioned);
```

Comment:
This logic is duplicated in `RendezvousCleanup` but that calls `OnPairingComplete`, which would mean it gets called again here.

It's unfortunate that I can't just re-use that, but would love some input...

Reply:
I think this is fine for now. Better than it was. Boris was talking earlier about tightening up some of the callbacks around pairing - the callbacks aren't that consistent and it's hard to tell precisely what they are supposed to represent. But our tooling also relies on some of these side effects, so a proper cleanup is a bit of a project. The plan was to look into this and clean it up a bit. Tracked here: https://github.com/project-chip/connectedhomeip/issues/15612